app/vmselect/promql/aggr.go

```diff
@@ -801,15 +801,43 @@ func avgValue(values []float64) float64 {
 }
 
 func medianValue(values []float64) float64 {
-	h := histogram.GetFast()
-	for _, v := range values {
-		if !math.IsNaN(v) {
-			h.Update(v)
-		}
+	return quantile(0.5, values)
+}
+
+func quantiles(phis []float64, values []float64) []float64 {
+	sort.Float64s(values)
+	res := make([]float64, len(phis))
+	for i, phi := range phis {
+		res[i] = quantileSorted(phi, values)
 	}
-	value := h.Quantile(0.5)
-	histogram.PutFast(h)
-	return value
+	return res
+}
+
+func quantile(phi float64, values []float64) float64 {
+	sort.Float64s(values)
```

Comment:
It looks like NaN values must be removed before sorting the rest of them. Otherwise sorting may work improperly, since the comparison with NaN always returns false.

Reply:
They should, indeed. In the code it is currently a caller's responsibility to remove NaN's. We could either make it clear that `values` argument should be without NaNs in the comments, or check it right in the function. wdyt?

Reply:
I think it would be easier from maintenance PoV to check this in the function. In this case the caller doesn't need to do this on its side. Additionally, I'd copy `values` before sorting them, since sometimes the caller doesn't expect that the `values` may be sorted (at least, `quantile_over_time` function).

Reply:
I'd recommend creating a pool for `*[]float64` values and using it for filtering for NaNs and sorting here. This way you will reduce the number of memory allocations to zero.

Reply:
please see update